### DISCUSSION TITLE
Folders for playlists

### DISCUSSION THREAD

**Discussion by ShaDream** [navidrome/navidrome#3371] [Ideas]
I want to create some themed playlists for myself and would like to separate them from others by moving them to a separate folder (in the application, not in the files on the disk.). Can this be implemented?

I think such a feature would be useful for those who like to categorize their library.

**Comment by kalenio** [comment]
This is a feature I would also appreciate. Similarly, a way to manually order playlists would be nice.

**Comment by thisolivier** [comment]
I'm very keen on this. I've done my own fork of the Amperfy client to facilitate this locally. Honestly it's closer to tags than a file system, and was light-weight to implement. 

## The basics
- Folders are DB entities with a many to many join table for playlist membership, and a name.
- Each folder can have many playlists
- A playlist can also be in many folders
- Playlists which appear at the top level are playlists with no folder membership
- Folders can contain folders, but each folder can only have one parent (many to one relationship).
- If a folder has no parent, it appears at the top level.

### Deleting Folders

Simply cascading database logic means when a folder is deleted, its playlists pop back to the root, but you could add UI to let the user choose whether to delete its contents too.

## Moving to Server Side
If we did adopt it on navidrome, we'd need to add the tables. Then I would build the folders as a separate endpoint, and add the folder membership array as a field to playlists. It would be backwards compatible with any clients that don't support folders since they would ignore the membership field, and never query for folders.

You could, if you wanted to force folder info on old clients, add prefixes to playlist names with their folder parent name by default, and skip that if the client declares support for folders, but it seems like overkill and fragile.

## The View Logic
In my implementation the view's presentational logic itself builds the folder tree, then maps the flat list of playlists into their folders. This is trivial to compute even for hundreds of playlists and dozens of folders. For larger libraries (the mind boggles), you might want to improve the query logic so we get the top level list, and the top level folders, but only query the contents of a folder once opened.

**Reply by thisolivier** [reply]
No replies on this, but I am going to build it.

**Reply by Ozidesk** [reply]
I'll keep an eye out for this if you end up building it.

**Reply by thisolivier** [reply]
I actually did already, it works great, but I've only exposed the views through Amperfy since I don't use the Navidrome web interface much. If someone wants to collaborate and work on the Navidrome UI part, we could ship it easy peasy.

**Reply by counterpaul** [reply]
This is very exciting! I don't use Amperfy so consider this a big vote for adding the feature server-side. I'd love to see developers of other clients be able to add this feature. Unfortunately I'm not a developer, so I can't offer to help.

**Reply by deluan** [reply]
Keep in mind that, for this to be adopted by Subsonic clients (and servers), it needs to be added to the API specification. There are already 2 open discussions for this, but nothing was decided yet:
https://github.com/opensubsonic/open-subsonic-api/discussions/14
https://github.com/opensubsonic/open-subsonic-api/discussions/75

While that is not approved and agreed by OpenSubsonic members, this won't be implemented.

**Reply by thisolivier** [reply]
I think what I'll do is wrap up my implementation and make it available as a fork, then dive into those discussions. I agree the goal is 1st class API support. 

Honestly, I could have an agent spit out the Navidrome UI components, but I'm just not motivated to make sure it's a good implementation (and I'd like to avoid LLM sins). What I could do is extract the product design from my Amperfy fork and submit THAT for consideration.

**Comment by thisolivier** [comment]
An update: The open-subsonic proposal is coming along. Another contributor had pipped me to the post and has since gone dark, but there's some interest and I think we'll get it over the line sometime soon. See https://github.com/opensubsonic/open-subsonic-api/pull/255 

I have a working fork of Navidrome which matches my proposed spec and I've been running for months, and an Amperfy client fork too, so once we get the open-subsonic feature I can submit a guy with a good amount of road testing under its belt.
